rclcpp/include/rclcpp/experimental/subscription_intra_process.hpp

```diff
@@ -134,6 +148,20 @@ class SubscriptionIntraProcess : public SubscriptionIntraProcessBase
     return buffer_->use_take_shared_method();
   }
 
+  void
+  provide_intra_process_message(const SerializedMessage & serialized_message)
```

Comment:
IMO, there's a minimal asymmetry in the implementation, that makes this extremely confusing:

- If both subscription and publisher type matches, we use this other [overload](https://github.com/ros2/rclcpp/blob/0577f390242de78eeec1ae75fda9be7193995e47/rclcpp/include/rclcpp/experimental/subscription_intra_process.hpp#L139), and that's also true if both publisher/subscription are using a serialized message (the only difference is that the other signature uses an `unique_ptr`).
- If the publisher is serialized and the subscription isn't, this method is used, which provides a non-serialized message from a serialized one.
- If the publisher isn't serialized and the subscription is, the [original overload]([overload](https://github.com/ros2/rclcpp/blob/0577f390242de78eeec1ae75fda9be7193995e47/rclcpp/include/rclcpp/experimental/subscription_intra_process.hpp#L139)) is used, and serialization is handled in `add_owned_msg_to_buffers`/`add_shared_msg_to_buffers`

We should either have methods clearly named here (`provide_serialized_intra_process_message_from_ros_message` and `provide_ros_intra_process_message_from_serialized_message`), or always handle serialization/deserialization in `add_owned_msg_to_buffers`/`add_shared_msg_to_buffers`.

Reply:
In my opinion, it makes sense to handle the serialization in the `add_*msg_to_buffers` or some central place.

From what I understand, the `provide_intra_process_message` functions should simply be overloaded with `SerializedMessages` and shouldn't do any further logic than message passing - pretty much exactly how the other functions work. The serialization/deserialization should be handled outside of this, i.e. in the `add` functions.

Reply:
I just played around a bit more with this and honestly, I don't know what the nicest solution for this is. 

I am generally in favor of handling the serialization in one place, ideally in the intra-process manager. In order to do that, we need a couple of functions in the subscription base:

This handle essentially returns the typesupport pointer which is required in order to run `rmw_serialize`.
```
const rosidl_type_support_t *
get_type_support_handle() const;
```
The implemenation of the intraprocess subscription can then do the SFINAE thing and either return the rosidl handle for `MessageT` if it's not a serialized message or return a `nullptr` otherwise.

The second message you'd need on the subscription base is to create/instantiate a default `MessageT` and return this as a `shared_ptr<void>` or `unique_ptr<void>` depending on the number of subscription.
```
std::shared_ptr<void> generate_message() const
{
  return std::make_shared<MessageT>();  // same for unique ptr
}
```
That returned pointer can then be filled by the serialization and passed back via `provide_intra_process_message`.

The `provide_intra_process_message` itself has to be overloaded with `shared_ptr<void>` and `unique_ptr<void>` whose implementation then casts these pointers back to the `MessageT` type and proceed as usual.
```
void provide_intra_process_message(const std::shared_ptr<void> message) {
  provide_intra_process_message(std::static_pointer_cast<const MessageT>(message));
}
```

That's quite a bit of flip-flop but doable. The alternatively, we could just invert the logic and do all the serialization inside the intraprocess subscription.

---

as for the SFINAE part, what I came up with which works nicely without too many empty implemented specializations is the following:

```
  const rosidl_message_type_support_t *
  get_type_support_handle() const override
  {
    return get_type_support_handle_impl<serialization_traits::is_serialized_message_class<MessageT>::value>();
  }

  template<bool is_serialized>
  const rosidl_message_type_support_t *
  get_type_support_handle_impl() const
  {
    return rosidl_typesupport_cpp::get_message_type_support_handle<MessageT>();
  }

  template<>
  const rosidl_message_type_support_t *
  get_type_support_handle_impl<true>() const
  {
    return nullptr;
  }
```

The serialization can then be conveniently be instantiated with:
```
rclcpp::SerializationBase serialization(get_type_support_handle());
```